File: app/src/main/java/com/uici/repertoire/MainActivity.java
package com.uici.repertoire;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.uici.repertoire.databinding.ActivityMainBinding;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements ContactAdapter.OnContactActionListener {

    private ActivityMainBinding binding;
    private ContactManager contactManager;
    private ContactAdapter contactAdapter;
    private List<Contact> contacts;

    private static final int REQUEST_CALL_PERMISSION = 1;
    private static final int REQUEST_SMS_PERMISSION = 2;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        contactManager = new ContactManager(this);
        contacts = new ArrayList<>();
        loadContacts();

        setupRecyclerView();
        setupFab();
    }

    @Override
    protected void onResume() {
        super.onResume();
        loadContacts();
    }

    private void setupRecyclerView() {
        contactAdapter = new ContactAdapter(contacts, this);
        binding.recyclerViewContacts.setLayoutManager(new LinearLayoutManager(this));
        binding.recyclerViewContacts.setAdapter(contactAdapter);
    }

    private void setupFab() {
        binding.fabAddContact.setOnClickListener(v -> {
            startActivity(new Intent(MainActivity.this, AddContactActivity.class));
        });
    }

    private void loadContacts() {
        contacts.clear();
        contacts.addAll(contactManager.getContacts());
        if (contactAdapter != null) {
            contactAdapter.notifyDataSetChanged();
        }
        updateEmptyState();
    }

    private void updateEmptyState() {
        if (contacts.isEmpty()) {
            binding.textViewEmpty.setVisibility(View.VISIBLE);
            binding.recyclerViewContacts.setVisibility(View.GONE);
        } else {
            binding.textViewEmpty.setVisibility(View.GONE);
            binding.recyclerViewContacts.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void onCallClick(Contact contact) {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CALL_PHONE)
                != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(
                    this,
                    new String[]{Manifest.permission.CALL_PHONE},
                    REQUEST_CALL_PERMISSION
            );
        } else {
            Intent intent = new Intent(Intent.ACTION_CALL);
            intent.setData(Uri.parse("tel:" + contact.getTelephone()));
            startActivity(intent);
        }
    }

    @Override
    public void onSmsClick(Contact contact) {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.SEND_SMS)
                != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(
                    this,
                    new String[]{Manifest.permission.SEND_SMS},
                    REQUEST_SMS_PERMISSION
            );
        } else {
            Intent intent = new Intent(Intent.ACTION_SENDTO);
            intent.setData(Uri.parse("smsto:" + contact.getTelephone()));
            startActivity(intent);
        }
    }

    @Override
    public void onLocationClick(Contact contact) {
        if (contact.getLatitude() != 0.0 && contact.getLongitude() != 0.0) {
            String uri = String.format("geo:%f,%f?q=%f,%f(%s)",
                    contact.getLatitude(), contact.getLongitude(),
                    contact.getLatitude(), contact.getLongitude(),
                    contact.getLieuHabitation());
            Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(uri));
            intent.setPackage("com.google.android.apps.maps");
            if (intent.resolveActivity(getPackageManager()) != null) {
                startActivity(intent);
            } else {
                Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(uri));
                startActivity(browserIntent);
            }
        } else {
            Toast.makeText(this, "Localisation non disponible", Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void onDeleteClick(Contact contact) {
        new AlertDialog.Builder(this)
                .setTitle("Supprimer le contact")
                .setMessage("Voulez-vous vraiment supprimer " + contact.getNom() + " " + contact.getPrenoms() + " ?")
                .setPositiveButton("Oui", (dialog, which) -> {
                    contactManager.deleteContact(contact.getId());
                    loadContacts();
                    Toast.makeText(MainActivity.this, "Contact supprimé", Toast.LENGTH_SHORT).show();
                })
                .setNegativeButton("Non", null)
                .show();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_CALL_PERMISSION || requestCode == REQUEST_SMS_PERMISSION) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                Toast.makeText(this, "Permission accordée", Toast.LENGTH_SHORT).show();
            } else {
                Toast.makeText(this, "Permission refusée", Toast.LENGTH_SHORT).show();
            }
        }
    }
}

File: app/src/main/java/com/uici/repertoire/MainActivity.kt
package com.uici.repertoire

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.uici.repertoire.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var contactManager: ContactManager
    private lateinit var contactAdapter: ContactAdapter
    private var contacts = mutableListOf<Contact>()

    companion object {
        private const val REQUEST_CALL_PERMISSION = 1
        private const val REQUEST_SMS_PERMISSION = 2
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        contactManager = ContactManager(this)
        loadContacts()

        setupRecyclerView()
        setupFab()
    }

    override fun onResume() {
        super.onResume()
        loadContacts()
    }

    private fun setupRecyclerView() {
        contactAdapter = ContactAdapter(
            contacts,
            onCallClick = { contact -> makeCall(contact) },
            onSmsClick = { contact -> sendSms(contact) },
            onLocationClick = { contact -> showLocation(contact) },
            onDeleteClick = { contact -> deleteContact(contact) }
        )

        binding.recyclerViewContacts.apply {
            layoutManager = LinearLayoutManager(this@MainActivity)
            adapter = contactAdapter
        }
    }

    private fun setupFab() {
        binding.fabAddContact.setOnClickListener {
            startActivity(Intent(this, AddContactActivity::class.java))
        }
    }

    private fun loadContacts() {
        contacts.clear()
        contacts.addAll(contactManager.getContacts())
        if (::contactAdapter.isInitialized) {
            contactAdapter.notifyDataSetChanged()
        }
        updateEmptyState()
    }

    private fun updateEmptyState() {
        if (contacts.isEmpty()) {
            binding.textViewEmpty.visibility = android.view.View.VISIBLE
            binding.recyclerViewContacts.visibility = android.view.View.GONE
        } else {
            binding.textViewEmpty.visibility = android.view.View.GONE
            binding.recyclerViewContacts.visibility = android.view.View.VISIBLE
        }
    }

    private fun makeCall(contact: Contact) {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CALL_PHONE)
            != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.CALL_PHONE),
                REQUEST_CALL_PERMISSION
            )
        } else {
            val intent = Intent(Intent.ACTION_CALL)
            intent.data = Uri.parse("tel:${contact.telephone}")
            startActivity(intent)
        }
    }

    private fun sendSms(contact: Contact) {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.SEND_SMS)
            != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.SEND_SMS),
                REQUEST_SMS_PERMISSION
            )
        } else {
            val intent = Intent(Intent.ACTION_SENDTO)
            intent.data = Uri.parse("smsto:${contact.telephone}")
            startActivity(intent)
        }
    }

    private fun showLocation(contact: Contact) {
        if (contact.latitude != 0.0 && contact.longitude != 0.0) {
            val uri = "geo:${contact.latitude},${contact.longitude}?q=${contact.latitude},${contact.longitude}(${contact.lieuHabitation})"
            val intent = Intent(Intent.ACTION_VIEW, Uri.parse(uri))
            intent.setPackage("com.google.android.apps.maps")
            if (intent.resolveActivity(packageManager) != null) {
                startActivity(intent)
            } else {
                val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(uri))
                startActivity(browserIntent)
            }
        } else {
            Toast.makeText(this, "Localisation non disponible", Toast.LENGTH_SHORT).show()
        }
    }

    private fun deleteContact(contact: Contact) {
        AlertDialog.Builder(this)
            .setTitle("Supprimer le contact")
            .setMessage("Voulez-vous vraiment supprimer ${contact.nom} ${contact.prenoms} ?")
            .setPositiveButton("Oui") { _, _ ->
                contactManager.deleteContact(contact.id)
                loadContacts()
                Toast.makeText(this, "Contact supprimé", Toast.LENGTH_SHORT).show()
            }
            .setNegativeButton("Non", null)
            .show()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            REQUEST_CALL_PERMISSION, REQUEST_SMS_PERMISSION -> {
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    Toast.makeText(this, "Permission accordée", Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(this, "Permission refusée", Toast.LENGTH_SHORT).show()
                }
            }
        }
    }
}
